### PR TITLE
Issue 305

### PR DESCRIPTION
**Pull request by b-gran**
Create independent APIs for
- Retrieving, creating, modifying, and deleting documents (what's currently in `operations.js`)
- Generating URIs from a model

### CODE REVIEW

File: src/express-restify-mongoose.js [hunk 1]
```diff
@@ -20,6 +23,16 @@ function getDefaults () {
   })
 }
 
+function ensureValueIsArray (value) {
+  if (_.isArray(value)) {
+    return value
+  }
+
+  return value
```
Comment (Zertz) [inline src/express-restify-mongoose.js]: This is a smart change! Just make this ternary a single line, for consistency.

File: src/RESTPathGenerator.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return ''
+  }
+
+  return paths.join('/').replace(/\/+/gi, '/')
+}
+module.exports.joinPathsAsURL = joinPathsAsURL
+
+class RESTPathGenerator {
+  constructor (apiPrefix = '', apiVersion = '', modelName) {
+    assert(typeof modelName === 'string', 'modelName must be a string')
```
Comment (Zertz) [inline src/RESTPathGenerator.js]: We should probably just throw here rather than use assert as a side effect

File: src/RESTPathGenerator.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
+/*
```
Comment (Zertz) [inline src/RESTPathGenerator.js]: This is awesome

**Comment by Zertz** [comment]
That's a heroic PR.

**Comment by b-gran** [comment]
Thanks! Locally, I'm working on updates to the operations API that so that each op returns a Promise -- it's still failing a couple tests though, but I should be able to sort it out in the morning.

By the way, great work on this library!

**Comment by Zertz** [comment]
I'm keeping an eye out on this, let me know if you have any questions!

File: src/api/database-operations/getItems.js [hunk 4]
```diff
@@ -1,48 +1,110 @@
...
+function getItems (queryOptions, mongooseContext, query) {
+  const buildQuery = getQueryBuilder(queryOptions)
 
-    options.contextFilter(model, req, (filteredContext) => {
-      buildQuery(filteredContext.find(), req._ermQueryOptions).then((items) => {
-        req.erm.result = items
-        req.erm.statusCode = 200
-
-        if (options.totalCountHeader && !req._ermQueryOptions['distinct']) {
-          options.contextFilter(model, req, (countFilteredContext) => {
```
Comment (b-gran) [inline src/api/database-operations/getItems.js]: In the old implementation, `contextFilter()` was called twice here -- once with the original request passed to the middleware, and then again with a _mutated_ request that has `items` attached and the status code set.

It is _theoretically_ possible that someone's `contextFilter()` depended on this behavior for this middleware (but I don't think it's likely). 

The new operations are "atomic" in the sense that they don't mutate the request until the middleware has finished running.

These changes still pass all of the tests we've got.

Comment (Zertz) [inline src/api/database-operations/getItems.js]: Good catch, while unlikely it could be a breaking change indeed. I wouldn't mind merging this in v5 as there are a couple other breaking changes I've had in mind for a little while. Notably, Node 0.10 and query operators (that jsonQueryParser you probably across) are going away.

**Comment by Zertz** [comment]
Sorry I kind of lost track of your PR, do you need me to check it out?

**Comment by b-gran** [comment]
I think we're fine for now. It's pretty close to being finished and at that
point I'd definitely appreciate it if you looked it over. I went down a bit
of a rabbit hole building a support library that I'm nearly finished with.

**Comment by Zertz** [comment]
Perfect, I'm going to have some time this weekend to look it over. Thanks again for the great work!

**Comment by Zertz** [comment]
@b-gran I added a few comments in the code, feel free to let me know if you disagree with them or have any questions :-)

**Comment by b-gran** [comment]
Okay great! Definitely appreciate all of your feedback, especially since there are so many changes to the guts of the library.

As for the comments back from October, I believe I incorporated all of the changes you suggested. Will double check, and if I missed any I'll make the changes now (they were all good ideas).

I checked for new comments after 19 Nov and didn't see any -- if you left any comments in the past day or so, would you mind sending me a link? I'm not super familiar with the PR UI and couldn't find a way to list all of the review & commit comments together in one place.

**Comment by b-gran** [comment]
BTW, here's my plan for finishing up this PR:

0. Finish [immutable-record](https://github.com/b-gran/immutable-record). It's related to ImmutableJS's [Record](http://facebook.github.io/immutable-js/docs/#/Record), but with a few key differences:
    * Backed by a plain JS object (rather than ImmutableJS's `Map`)
    * Fields are enumerable so `Object.keys(immutableRecord)` works just like it would on a plain object
    * Possible to make fields required/optional and provide complex validation functions
    * Possible to implement custom getters so if e.g. you want to store a `ModelQuery` in the Record, you can return `query.toConstructor()()` so the underlying query can't be mutated
1. Have `ERMOperation` extend from [immutable-record](https://github.com/b-gran/immutable-record).
2. Implement an `OperationContext` (also extends from `immutable-record`) class that wraps around both `ModelQuery`s and `Document`s and provides an API for accessing them.
3. Implement all of the middleware operations in `src/middleware/` as `APIMethod`s.
4. Have the main ERM function expose `APIMethod`s to consumers. We still need to decide how we want to do this.
5. Have some people with extensive ERM applications in the wild test this out, to make sure there aren't edge cases we missed with tests (e.g. someone's `contextFilter()` or `preMiddleware()` hook mutates the request in some strange way). I'll do this myself with our applications, but mine deliberately avoid mutating the request so I'd be more interested in seeing how it copes with others' applications.

Here are some options I've thought of for exposing the `APIMethod`s at the top level:

* Some kind of `APIWrapper` just sits at the top level of the ERM object. It gets initialized when the ERM function is called, and has all of the functions in `src/api/` and `src/middleware/` (once they're converted to `APIMethods`, of course). The `APIMethod`s are documented on the website.
* A top-level `erm.api()` function that explicitly initializes an `APIWrapper`
* The `APIMethods` themselves sit at the top level of the ERM object, so you could do `erm.prepareQuery(someObj)`
* Chained together `APIMethods` with names like `getAllDocuments`, `updateSingleDocument`, `patchSingleDocument`, etc. are available at the top level of the ERM object. They each return Promises that resolve to `ERMOperations`. These chained methods correspond roughly to the middleware chains we currently have in `express-restify-mongoose.js` (e.g. `app.get(restPaths.allDocuments, prepareQuery, ...`)

It might be nice to combine a couple of these options: expose the chained `APIMethod`s, and then also have `erm.api()` for people who want to use the individual methods. 

Exposing the individual `APIMethods` does increase the support burden for us though -- if we have e.g. `prepareQuery()` in the docs, we won't be able to change its implementation at will.

**Comment by Zertz** [comment]
You should see review comments right above my previous comment or in the **Files changed** tab.

**Comment by b-gran** [comment]
Hm that's strange, not seeing anything in "Files changed." 
Is it possible they're hidden somehow, or I don't have a permission to view them?

**Comment by Zertz** [comment]
Oops, first time using the GitHub review tools. You should be able to see my comments now :-)

File: src/RESTPathGenerator.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    this._apiVersion = apiVersion
+    this._modelName = modelName
+
+    this._basePath = joinPathsAsURL(
+      this._apiPrefix + this._apiVersion,
+      this._modelName
+    )
+
+    // Is a document id already in the base path?
+    const basePathContainsDocumentId = this._basePath.indexOf('/:id') !== -1
```
Comment (b-gran) [inline src/RESTPathGenerator.js]: Taking it out works for me. v5 can have breaking changes, right?

File: src/api/database-operations/deleteItem.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+              return resolve(deleteSuccess(state))
+            })
+            .catch(err => reject(err))
+        }
+      )
+    })
+  } else {
+    // Not using findOneAndRemove(), so just remove the document directly.
+    if (!state.document) {
+      return Promise.reject(new Error('No document found'))
```
Comment (b-gran) [inline src/api/database-operations/deleteItem.js]: Agreed

File: src/api/database-operations/deleteItem.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    return state.document.remove()
+      .then(_.constant(deleteSuccess(state)))
```
Comment (b-gran) [inline src/api/database-operations/deleteItem.js]: Good catch, this was a mistake.

File: src/api/database-operations/modifyObject.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+const APIMethod = require('../../APIMethod')
+const moredots = require('moredots')
+const _ = require('lodash')
+const http = require('http')
+
+/**
+ * Given a document and the document's model, depopulate any populated fields in the document.
+ * Does NOT mutate the populated document.
+ */
+function depopulate (model, populatedDoc) {
```
Comment (b-gran) [inline src/api/database-operations/modifyObject.js]: 😄

File: src/operations.js [hunk 9]
```diff
@@ -1,236 +1,13 @@
...
-      req.erm.document.save().then((item) => model.populate(item, req._ermQueryOptions.populate || [])).then((item) => {
-        req.erm.result = item
-        req.erm.statusCode = 200
-
-        next()
-      }, errorHandler(req, res, next))
-    }
-  }
+module.exports = function (ermInstance) {
+  const getItems = require('./api/database-operations/getItems').getMiddleware(ermInstance)
```
Comment (b-gran) [inline src/operations.js]: Agreed

**Comment by Zertz** [comment]
Looking good! I can give you a hand getting these tests to pass if you want me to.

**Comment by b-gran** [comment]
That would be great. It looks like they're all to do with error message format changes?

**Comment by b-gran** [comment]
Okay, it's fixed 😉

**Comment by Zertz** [comment]
Don't worry about the failures in 0.10, I'll drop it and 0.12 in a separate PR.

**Comment by Zertz** [comment]
You can go ahead and merge master into your branch, I cherry-picked some of your fixes so you shouldn't get any merge conflicts. CI should be green afterwards 😃

File: src/buildQuery.js [hunk 10]
```diff
@@ -1,50 +1,62 @@
...
+ * partial query with local (request-specific) query options.
+ *
+ * @param {Object} globalOptions - ERM-instance query options
+ * @param {Number} globalOptions.limit -
+ * @param {String} globalOptions.readPreference -
+ * @param {Boolean} globalOptions.lean -
+ *
+ * @return {function(ModelQuery, Object): Promise}
+ */
+module.exports = function (globalOptions) {
```
Comment (Zertz) [inline src/buildQuery.js]: Don't worry about it, we can refactor it out later.

File: src/RESTPathGenerator.js (hunk 5 above)
Comment (Zertz) [inline src/RESTPathGenerator.js]: Yes.

**Comment by Zertz** [comment]
Let me know if you need anything else from me :)

**Comment by b-gran** [comment]
Just realized the tests from #317 didn't make it in -- just added them.

**Comment by b-gran** [comment]
Hey @Zertz, what do you think about moving to ES6 modules? The tooling supports it, but it means modifying the tests (which have nested `require()` calls, so can't generally be converted) like this:

```
const restify = require('../express-restify-mongoose').default
```

**Comment by b-gran** [comment]
So I looked at the test that's failing and in 4.8.0 mongoose added support for selecting fields by array (the way the test attempts).

Should I update the test to document that new feature?

**Comment by OmgImAlexis** [comment]
@b-gran is this still being worked on?

**Comment by Zertz** [comment]
I love the idea here but it seems like work has stalled and quite a few merge conflicts have risen and the scope of changes is quite large. Closing for now, let me know if there is interest in picking up the work.
